File: backend/src/main/java/com/example/backend/service/AdminService.java
package com.example.backend.service;

import com.example.backend.dto.request.AdminRegisterRequest;
import com.example.backend.dto.request.AdminUpdateRequest;
import com.example.backend.dto.response.AdminResponse;
import com.example.backend.dto.response.RegisterResponse;
import com.example.backend.entity.enums.AdminStatus;
import com.example.backend.entity.enums.Role;
import org.springframework.data.domain.Page;

/**
 * Service for Admin management
 */
public interface AdminService {

    /**
     * Register a new admin/staff
     * @param request Admin registration request
     * @return Registration response
     */
    RegisterResponse registerAdmin(AdminRegisterRequest request);

    /**
     * Get admin information by keycloakId
     * @param keycloakId Keycloak user ID
     * @return Admin response
     */
    AdminResponse getAdminInfo(String keycloakId);

    /**
     * Get admin information by userId
     * @param userId User ID (String UUID)
     * @return Admin response
     */
    AdminResponse getAdminById(String userId);

    /**
     * Update admin profile
     * @param keycloakId Keycloak user ID
     * @param request Update request
     * @return Updated admin response
     */
    AdminResponse updateProfile(String keycloakId, AdminUpdateRequest request);

    /**
     * Update admin status (by super admin)
     * @param userId User ID (String UUID)
     * @param status New status
     * @return Updated admin response
     */
    AdminResponse updateStatus(String userId, AdminStatus status);

    /**
     * Activate/Deactivate admin account
     * @param userId User ID (String UUID)
     * @param active Active status
     * @return Updated admin response
     */
    AdminResponse setActive(String userId, boolean active);

    /**
     * Update admin role (by super admin)
     * @param userId User ID (String UUID)
     * @param role New role
     * @return Updated admin response
     */
    AdminResponse updateRole(String userId, Role role);

    /**
     * Get all admins with pagination and filtering
     * @param page Page number (0-based)
     * @param size Page size
     * @param role Filter by role (optional)
     * @param status Filter by status (optional)
     * @return Page of admin responses
     */
    Page<AdminResponse> getAllAdmins(int page, int size, Role role, AdminStatus status);
}

File: website/fe_admin/app/service/adminService.ts
import axiosInstance from '../config/axios';
import type { ApiResponse, PageResponse } from './types';

// Types
export interface AdminRegisterRequest {
  username: string;
  email: string;
  password: string;
  fullName: string;
  phoneNumber: string;
  role: 'ROLE_SUPER_ADMIN' | 'ROLE_MODERATOR' | 'ROLE_STAFF';
  avatarUrl?: string;
}

export interface AdminUpdateRequest {
  fullName?: string;
  email?: string;
  phoneNumber?: string;
  avatarUrl?: string;
}

export interface AdminResponse {
  // User fields
  userId: string;
  keycloakId: string;
  username: string;
  email: string;
  phoneNumber: string;
  fullName: string;
  avatarUrl: string;
  active: boolean;

  // Admin-specific fields
  role: string;
  status: string; // PENDING_APPROVAL, ACTIVE, INACTIVE
  lastLoginIp: string;

  // Metadata
  createdAt: string;
  updatedAt: string;
}

export interface RegisterResponse {
  userId: string;
  username: string;
  email: string;
  message: string;
}

// Admin Service
class AdminService {
  private handleError(error: any): Error {
    return new Error(error.response?.data?.message || error.message || 'An unexpected error occurred');
  }

  /**
   * Register new admin/staff (SUPER_ADMIN only)
   */
  async registerAdmin(request: AdminRegisterRequest): Promise<RegisterResponse> {
    try {
      const { data } = await axiosInstance.post<ApiResponse<RegisterResponse>>('/admins/register', request);
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Get current admin profile
   */
  async getCurrentAdmin(): Promise<AdminResponse> {
    try {
      const { data } = await axiosInstance.get<ApiResponse<AdminResponse>>('/admins/me');
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Get admin by ID (super admin only)
   */
  async getAdminById(userId: string): Promise<AdminResponse> {
    try {
      const { data } = await axiosInstance.get<ApiResponse<AdminResponse>>(`/admins/${userId}`);
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Update current admin profile
   */
  async updateProfile(request: AdminUpdateRequest): Promise<AdminResponse> {
    try {
      const { data } = await axiosInstance.put<ApiResponse<AdminResponse>>('/admins/me', request);
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Get all admins with pagination (super admin only)
   */
  async getAllAdmins(
    page: number = 0,
    size: number = 20,
    role?: string,
    status?: string
  ): Promise<PageResponse<AdminResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        size: size.toString(),
      });
      
      if (role && role !== 'ALL') params.append('role', role);
      if (status && status !== 'ALL') params.append('status', status);

      const { data } = await axiosInstance.get<ApiResponse<PageResponse<AdminResponse>>>(
        `/admins?${params.toString()}`
      );
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Approve admin (set to ACTIVE status)
   */
  async approveAdmin(userId: string): Promise<AdminResponse> {
    try {
      const { data } = await axiosInstance.patch<ApiResponse<AdminResponse>>(`/admins/${userId}/approve`);
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Suspend admin account
   */
  async suspendAdmin(userId: string, reason?: string): Promise<AdminResponse> {
    try {
      const params = reason ? `?reason=${encodeURIComponent(reason)}` : '';
      const { data } = await axiosInstance.patch<ApiResponse<AdminResponse>>(`/admins/${userId}/suspend${params}`);
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Activate admin account
   */
  async activateAdmin(userId: string): Promise<AdminResponse> {
    try {
      const { data } = await axiosInstance.patch<ApiResponse<AdminResponse>>(`/admins/${userId}/activate`);
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Update admin role (super admin only)
   */
  async updateAdminRole(userId: string, role: string): Promise<AdminResponse> {
    try {
      const { data } = await axiosInstance.patch<ApiResponse<AdminResponse>>(
        `/admins/${userId}/role?role=${role}`
      );
      return data.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Update admin status (super admin only)
   */
  async updateAdminStatus(userId: string, status: string): Promise<AdminResponse> {
    try {
      // Use existing endpoints based on status
      if (status === 'ACTIVE') {
        return await this.approveAdmin(userId);
      } else if (status === 'INACTIVE') {
        return await this.suspendAdmin(userId);
      }
      throw new Error('Invalid status');
    } catch (error) {
      throw this.handleError(error);
    }
  }
}

// Export singleton instance
export default new AdminService();
